### ISSUE TITLE
ReflectionException: Class not found

### ISSUE THREAD

**Issue by ivansrdic** [api-platform/core#3349]
**API Platform version(s) affected**: 2.5.4 (from commit https://github.com/api-platform/core/commit/baf054234eff3b9276b6783efb825bff7099d5c9)

**Description** / **How to reproduce**  
Loading the api docs causes several "class not found" ReflectionExceptions.
![Screenshot from 2020-01-20 13-53-51](https://user-images.githubusercontent.com/7668873/72727966-86212200-3b8c-11ea-967f-d9cdf07f7d35.png)
Possibly related to https://github.com/api-platform/core/issues/3344

**Possible Solution**  
Seems to be related to `src/JsonSchema/TypeFactory:102`. If the check in changed from and to or, everything works (working example below)
```
        if ($this->isResourceClass($className) || true !== $readableLink) {
            return [
                'type' => 'string',
                'format' => 'iri-reference',
            ];
        }
```

Unfortunately I don't know the internals enough to know if this breaks anything.

**Comment by soyuka** [comment]
Definitely related. Could you contribute your solution?

**Comment by soyuka** [comment]
btw it looks like this class really doesn't exist, you have a bad import.

**Comment by ivansrdic** [comment]
I've checked, import is correct. The error message is gone if in the related trait I remove the `use` for the EventDispatcherInterface statement and replace the class references with the FQN. But the error appears with all used traits that have properties that are used in ApiResource entities. I guess this means that the class resolver has a bug somewhere.

Example of when it throws an exception
```
<?php

namespace App\DependencyInjection;

use \Symfony\Component\EventDispatcher\EventDispatcherInterface;

trait EventDispatcherAwareTrait
{

    /**
     * @var EventDispatcherInterface
     */
    protected $eventDispatcher;

    /**
     * @inheritDoc
     */
    public function setEventDispatcher(EventDispatcherInterface $eventDispatcher)
    {
        $this->eventDispatcher = $eventDispatcher;
    }
}
```

No exception after this change
```
<?php

namespace App\DependencyInjection;

trait EventDispatcherAwareTrait
{

    /**
     * @var \Symfony\Component\EventDispatcher\EventDispatcherInterface
     */
    protected $eventDispatcher;

    /**
     * @inheritDoc
     */
    public function setEventDispatcher(\Symfony\Component\EventDispatcher\EventDispatcherInterface $eventDispatcher)
    {
        $this->eventDispatcher = $eventDispatcher;
    }
}
```

**Comment by ivansrdic** [comment]
btw https://github.com/api-platform/core/pull/3350 causes another error
![Screenshot from 2020-01-20 15-22-29](https://user-images.githubusercontent.com/7668873/72733725-b7075400-3b98-11ea-82e4-f34a98c1e4d0.png)

**Comment by soyuka** [comment]
Not this namespace, the one you have in your Entity: `App\Entity\Custom\InternalNote\EventDispatcherInterface` should be  `\Symfony\Component\EventDispatcher\EventDispatcherInterface` in the error. I can not reproduce this with the same code on my end. 

Now trying to reproduce the `$ref` but same...

**Comment by soyuka** [comment]
see https://github.com/soyuka/self-referencing-metadata-api-platform for the reproduction

**Comment by ivansrdic** [comment]
@soyuka That's the point - I don't have that property defined in the Entity. It is defined in the trait I pasted above and used in the Entity. 

Let me be more verbose:
I have an entity `\App\Entity\Custom\InternalNote\InternalNote`
```
<?php

namespace App\Entity\Custom\InternalNote;

use App\Contract\Entity as EntityContract;

class InternalNote implements EntityContract\Audit\AuditableInterface
{
    use EntityContract\Audit\AuditableTrait;
    ...
}
```

`AuditableTrait` content is
```
<?php

namespace App\Contract\Entity\Audit;

use App\DependencyInjection\EventDispatcherAwareTrait;

trait AuditableTrait
{
    use EventDispatcherAwareTrait;
    ...
}
```

`EventDispatcherAwareTrait` content is
```
<?php

namespace App\DependencyInjection;

use \Symfony\Component\EventDispatcher\EventDispatcherInterface;

trait EventDispatcherAwareTrait
{

    /**
     * @var EventDispatcherInterface
     */
    protected $eventDispatcher;

    /**
     * @inheritDoc
     */
    public function setEventDispatcher(EventDispatcherInterface $eventDispatcher)
    {
        $this->eventDispatcher = $eventDispatcher;
    }
}
```

The entity does not redeclare the `$eventDispatcher` property nor any other property from that trait "chain".

**Comment by soyuka** [comment]
nevermind I got to reproduce the bug! thanks

**Comment by soyuka** [comment]
So, I tracked the bugs to php documentor I'll open issues their regarding these issues.

In the mean time you should use the FQDN in the Trait property... I'll try to find a quicker solution to these problems.

**Comment by bastnic** [comment]
FYi, I've got a funny one, because of that error in a vendor `@return Returns blablablab`:

https://github.com/KnpLabs/DoctrineBehaviors/blob/9cc036ee32483c306fb32b4efbf33014ddfb654f/src/Model/Translatable/TranslatableMethods.php#L149

I got errors like 

```
In SchemaFactory.php line 216:
                                                                  
  Class Lib\Core\Entity\Catalog\Reference\Returns does not exist  
```

Because in a class in namespace `Lib\Core\Entity\Catalog\Reference` i've a `use TranslatableTrait` which use `TranslatableMethods`

**Comment by GPMathieu** [comment]
@soyuka Do you have some news about this issue ?
Because i tried to used https://github.com/KnpLabs/DoctrineBehaviors/blob/master/docs/translatable.md and i have : 
Class App\Entity\TranslationInterface does not exist

**Comment by azribilel** [comment]
i'm still use version 2.5.3 cause of this issue :(

**Comment by soyuka** [comment]
We need to fix https://github.com/symfony/symfony/issues/28732  :/

**Comment by vic-blt** [comment]
> @soyuka Do you have some news about this issue ?
> Because i tried to used https://github.com/KnpLabs/DoctrineBehaviors/blob/master/docs/translatable.md and i have :
> Class App\Entity\TranslationInterface does not exist
> 
> Symfony : 5.0.9
> API Platform : 2.5.6
> 
> ![error](https://user-images.githubusercontent.com/60045205/83753260-430ea980-a66a-11ea-8c09-64e0c49a1ce8.png)

Same issue.
Symfony: 5.1.7
API Platform: 2.5.7

**Comment by cnizzardini** [comment]
This arrises if you you do something like this:

`@ODM\ReferenceOne(targetDocument=Services::class, storeAs="id" , inversedBy="system")` and the class doesn't exist. The error should be more exact if possible.

**Comment by DanidCH** [comment]
> > @soyuka Do you have some news about this issue ?
> > Because i tried to used https://github.com/KnpLabs/DoctrineBehaviors/blob/master/docs/translatable.md and i have :
> > Class App\Entity\TranslationInterface does not exist
> > Symfony : 5.0.9
> > API Platform : 2.5.6
> > ![error](https://user-images.githubusercontent.com/60045205/83753260-430ea980-a66a-11ea-8c09-64e0c49a1ce8.png)
> 
> Same issue.
> Symfony: 5.1.7
> API Platform: 2.5.7
> 
> **EDIT**:
> 
> As @soyuka mentioned
> 
> > In the mean time you should use the FQDN in the Trait property [...]
> 
> For `knplabs/doctrine-behaviors` bundle, you have to override classes below and remove the `@var` or use the FQDN for the `@var` :
> 
> * `TranslatablePropertiesTrait`
>   https://github.com/KnpLabs/DoctrineBehaviors/blob/2fb26d94aee366a267d9b31791f2f799b691bb8c/src/Model/Translatable/TranslatablePropertiesTrait.php#L13
>   https://github.com/KnpLabs/DoctrineBehaviors/blob/2fb26d94aee366a267d9b31791f2f799b691bb8c/src/Model/Translatable/TranslatablePropertiesTrait.php#L19
> * `TranslatableMethodsTrait`
>   https://github.com/KnpLabs/DoctrineBehaviors/blob/2fb26d94aee366a267d9b31791f2f799b691bb8c/src/Model/Translatable/TranslatableMethodsTrait.php#L161

@vic-blt: tried this unsuccessfully, would you have more details or an example on your work around?

**Comment by vic-blt** [comment]
> tried this unsuccessfully, would you have more details or an example on your work around?

Unfortunately no because I opted in for another API solution.
I'm currently using API Gateway of AWS.

**Comment by SherinBloemendaal** [comment]
Is there a solution? Swagger is currently not working so pretty annoying.

**Comment by soullivaneuh** [comment]
I am not sure if the error is "normal" or not. Apparently, it comes from real phpdoc mistakes like `@return Returns something` without the type.

However, it would be a very nice addition if api-platform can handle the error properly and dispatch useful information like the related not working property.

Currently, I have to put some `dump` on `SchemaFactory::buildPropertySchema` to get the information and provide the things.
That method slow down a lot the debug process.

**Comment by soullivaneuh** [comment]
FYI, I update `symfony/property-info` to v4.4.20, including the apparently related fix: https://github.com/symfony/symfony/commit/1b358fef04861920fa2550db5d23c1b3f6bc3a5f

However, it result to the same error.

**Comment by soullivaneuh** [comment]
Still a `symfony/property-info` related bug, please see: https://github.com/symfony/symfony/issues/40839

Maybe it's multiple bug, but not related to api-platform.

The only way to solve the api-platform related issue is to put a cleaner error management like I suggested in https://github.com/api-platform/core/issues/3349#issuecomment-814209613, or so nothing. :-)

**Comment by soullivaneuh** [comment]
For information, the last error I have is related to knp-behavior v2 trait:

```php
use ApiPlatform\Core\Annotation\ApiResource;
use Doctrine\ORM\Mapping as ORM;
use Knp\DoctrineBehaviors\Contract\Entity\TranslatableInterface;
use Knp\DoctrineBehaviors\Model\Translatable\TranslatableTrait;

/**
 * @ORM\Table
 * @ORM\Entity
 * @ApiResource(
 *     collectionOperations={
 *         "get",
 *     },
 *     itemOperations={
 *         "get",
 *     },
 * )
 */
class Accounting implements TranslatableInterface
{
    use TranslatableTrait;
// ...
```

For me, all is related to the property-info component.

For that case, I answered on a closed bug report that exactly match my case with the supposed to be fixed releases: https://github.com/symfony/symfony/issues/36997#issuecomment-822572640

**Comment by soullivaneuh** [comment]
I proposed a fix: https://github.com/symfony/symfony/pull/40874

However, there is a workaround for the `Knp\DoctrineBehaviors` case only, import directly the trait imported by the main one.

In a nutshell:

```diff
use Knp\DoctrineBehaviors\Model\Translatable\TranslatableMethodsTrait;
use Knp\DoctrineBehaviors\Model\Translatable\TranslatablePropertiesTrait;

class Accounting implements TranslatableInterface
{
    use TranslatablePropertiesTrait;
    use TranslatableMethodsTrait;
}
```

Hope it helps!

**Comment by angelomelonas** [comment]
I would just like to mention that I ran into this issue with `TranslatableTrait` running Symfony 6.2.*. The suggested workaround from @soullivaneuh did the trick though. However, I also had an issue with another trait that contained a reference to a value object. The value object in the trait is referenced as follows:
```
    /** @var array<string, ValueObjectData> */
    private array $valueObjectData = [];
``` 
Removing or replacing the PHP doc with   `/** @var array<mixed> */` resolves this issue, but breaks some PHPStan rules. I ended up solving this by moving the  `ValueObjectData` class to a directory outside of the one the trait is located in.
